File: leetcode/295-find-median-from-data-stream/main.go
package main

import (
	"container/heap"
	"fmt"
)

/*
	3rd approach: 2 heaps
	- minheap for right half
	- maxheap for left half
	- therefore the mean will always be the max of maxheap and the min of minheap
	- https://leetcode.com/articles/find-median-from-data-stream/

	TimeAddNum() 				O(5*logn)
	Time FindMedian() 	O(1)
	132 ms, faster than 80.65%
*/

type MedianFinder struct {
	maxHeap *MaxHeap
	minHeap *MinHeap
}

/** initialize your data structure here. */
func Constructor() MedianFinder {
	left := &MaxHeap{}
	right := &MinHeap{}
	return MedianFinder{left, right}
}

func (this *MedianFinder) AddNum(num int) {
	heap.Push(this.minHeap, num)
	m := heap.Pop(this.minHeap)
	heap.Push(this.maxHeap, m)
	if this.minHeap.Len() < this.maxHeap.Len() {
		n := heap.Pop(this.maxHeap)
		heap.Push(this.minHeap, n)
	}
}

func (this *MedianFinder) FindMedian() float64 {
	fmt.Println(this.maxHeap)
	fmt.Println(this.minHeap)
	if this.minHeap.Len() != this.maxHeap.Len() {
		return float64((*this.minHeap)[0])
	}
	return (float64((*this.minHeap)[0]) + float64((*this.maxHeap)[0])) / 2.0
}

// min heap
type MinHeap []int

func (h MinHeap) Len() int            { return len(h) }
func (h MinHeap) Less(i, j int) bool  { return h[i] < h[j] }
func (h MinHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *MinHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// max heap
type MaxHeap []int

func (h MaxHeap) Len() int            { return len(h) }
func (h MaxHeap) Less(i, j int) bool  { return h[i] > h[j] }
func (h MaxHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func main() {

	obj := Constructor()
	obj.AddNum(1)
	obj.AddNum(2)
	fmt.Println(obj.FindMedian())

	obj = Constructor()
	obj.AddNum(5)
	obj.AddNum(3)
	obj.AddNum(4)
	obj.AddNum(3)
	obj.AddNum(1)
	obj.AddNum(2)
	obj.AddNum(6)
	fmt.Println(obj.FindMedian())
}

File: leetcode/295-find-median-from-data-stream/main.py
import heapq

"""
    1st approach: heap
    - LTE

    2nd approach: BST
    - LTE
"""

"""
    3rd approach: 2 heaps
    - maxheap for left half
    - minheap for right half
    - therefore the mean will always be the max of maxheap and the min of minheap
    - https://leetcode.com/articles/find-median-from-data-stream/

    always maintain the maxheap <= minheap in size

    308 ms, faster than 55.77%
"""


class MedianFinder(object):

    def __init__(self):
        # for left half
        self.maxheap = []
        # for right half
        self.minheap = []

    def addNum(self, num):
        """
        O(3*logn)
        """
        if len(self.maxheap) == len(self.minheap):
            heapq.heappush(self.maxheap, -num)
            largest = heapq.heappop(self.maxheap)
            heapq.heappush(self.minheap, -largest)
        else:
            heapq.heappush(self.minheap, num)
            least = heapq.heappop(self.minheap)
            heapq.heappush(self.maxheap, -least)

    def findMedian(self):
        """
        O(1)
        """
        if len(self.maxheap) != len(self.minheap):
            return self.minheap[0]
        return (-self.maxheap[0] + self.minheap[0])/2.0


obj = MedianFinder()
obj.addNum(1)
print(obj.maxheap)
print(obj.minheap)
print(obj.findMedian())

print("---")

obj = MedianFinder()
obj.addNum(1)
obj.addNum(2)
print(obj.maxheap)
print(obj.minheap)
print(obj.findMedian())

print("---")

obj = MedianFinder()
obj.addNum(1)
obj.addNum(3)
obj.addNum(2)
print(obj.maxheap)
print(obj.minheap)
print(obj.findMedian())

print("---")

obj = MedianFinder()
obj.addNum(5)
obj.addNum(3)
obj.addNum(4)
obj.addNum(3)
obj.addNum(1)
obj.addNum(2)
obj.addNum(6)
print(obj.maxheap)
print(obj.minheap)
print(obj.findMedian())

print("========================================")

"""
    3rd approach2: 2 heaps
    - maxheap for left half
    - minheap for right half
    - therefore the mean will always be the max of maxheap and the min of minheap
    - https://leetcode.com/articles/find-median-from-data-stream/

    always maintain the maxheap >= minheap in size

    308 ms, faster than 55.77%
"""


class MedianFinder(object):

    def __init__(self):
        self.maxheap = []
        self.minheap = []

    def addNum(self, num):
        """
        O(logn)
        """
        if len(self.maxheap) == len(self.minheap):
            # push to right, pop right, push to left such that left will be larger
            heapq.heappush(self.minheap, num)
            leastOnRight = heapq.heappop(self.minheap)
            heapq.heappush(self.maxheap, -leastOnRight)
        else:
            # push to left, pop left, push to right such that left len == right len
            heapq.heappush(self.maxheap, -num)
            largestOnLeft = heapq.heappop(self.maxheap)
            heapq.heappush(self.minheap, -largestOnLeft)

    def findMedian(self):
        """
        O(1)
        """
        if len(self.maxheap) != len(self.minheap):
            return -self.maxheap[0]
        return (-self.maxheap[0] + self.minheap[0])/2.0


obj = MedianFinder()
obj.addNum(1)
print(obj.maxheap)
print(obj.minheap)
print(obj.findMedian())

print("---")

obj = MedianFinder()
obj.addNum(1)
obj.addNum(2)
print(obj.maxheap)
print(obj.minheap)
print(obj.findMedian())

print("---")

obj = MedianFinder()
obj.addNum(1)
obj.addNum(3)
obj.addNum(2)
print(obj.maxheap)
print(obj.minheap)
print(obj.findMedian())

print("---")

obj = MedianFinder()
obj.addNum(5)
obj.addNum(3)
obj.addNum(4)
obj.addNum(3)
obj.addNum(1)
obj.addNum(2)
obj.addNum(6)
print(obj.maxheap)
print(obj.minheap)
print(obj.findMedian())

print("========================================")

"""
    4th approach: upper bound binary search
    - add the number in correct position
    - find median from the half of the array

    Time    O(n) because list.insert() takes O(n)
    Space   O(n)
    336 ms, faster than 45.33%
"""


class MedianFinder(object):

    def __init__(self):
        self.sorted = []

    def addNum(self, num):
        """
        O(logn)
        """
        i = self._upperBSearch(num)
        self.sorted.insert(i, num)

    def _upperBSearch(self, target):
        left = 0
        right = len(self.sorted)
        while left < right:
            mid = (left + right)//2
            if target >= self.sorted[mid]:
                left = mid + 1
            else:
                right = mid
        return left

    def findMedian(self):
        """
        O(1)
        """
        if len(self.sorted) % 2 == 0:
            mid = (len(self.sorted)-1)/2
            left = self.sorted[mid]
            right = self.sorted[mid+1]
            return (left + right)/2.0
        mid = (len(self.sorted)-1)/2
        return self.sorted[mid]*1.0
